### ISSUE TITLE
Extracting cross-references from ontology and saving to a tsv

### ISSUE THREAD

**Issue by dhimmel** [owlcollab/owltools#128]
What command can I use to extract the value for `xref` tags from terms and save the results to a tsv? The tsv would preferably contain the following columns: term_id, term_name, xref. More generally, I am interested in extracting tag-value pairs from ontologies using owltools.

**Comment by hdietze** [comment]
Hello,
currently there is no general purpose command-line command to create a TSV (for xrefs or other OBO tags).
For the GeneOntology, we use the `--external-mappings-files` (`CommandRunner#createExternalMappings` is the underlying implementation) to generate TSV like files for a specific subsets of xrefs. But that command is quite specific for the GO use-case.

Under the hood the method makes use of the OWLGraphWrapper and its wrapper methods to access the appropriate OWL equivalent for an OBO-Format tag (such as `xref`).

There are two potential approaches:
 1. Write your own (generic?) method to dump out TSVs. (Use the existing code as example)
 2. OBO format (instead of the OWL-API) to extract the values. 
     See `org.obolibrary.oboformat.parser.OBOFormatParser` and  `org.obolibrary.oboformat.model.OBODoc` are a good start iwhen dealing with plain OBO.

**Comment by cmungall** [comment]
OBODoc is really intended only as an intermediate model on the way to OWL; there's no official support for using it as an API in its own right. But others have done this happily.

You could also do this via a SPARQL query

```
prefix oio: <http://www.geneontology.org/formats/oboInOwl#>
prefix owl: <http://www.w3.org/2002/07/owl#>
prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>

SELECT ?x ?l ?y WHERE {
  ?x a owl:Class ;
     rdfs:label ?l ;
     oio:hasDbXref ?y
}
```

Many ontologies have a sparql folder for these kinds of queries
https://github.com/obophenotype/uberon/tree/master/sparql

**Comment by dhimmel** [comment]
I've tried the SPARQL method without much luck. See the [notebook here](https://github.com/dhimmel/uberon/blob/912c81125d59ca4d94b4c7ed1831693f800872f3/sparql.ipynb).

I changed `http://www.geneontology.org/formats/oboInOwl#` to `ftp://ftp.geneontology.org/go/www/formats/oboInOwl#` because the former was broken.

I didn't know what url to use for a SPARQL endpoint or whether I needed to create my own. I unsuccessfully tried the endpoint (`http://sparql.hegroup.org/sparql`) mentioned on [Uberon's wikipedia page](https://en.wikipedia.org/wiki/Uberon). Can this SPARQL method read directly from an owl file?

**Comment by cmungall** [comment]
Yes, I typically use `arq` which I think is pre-bundled with apache jena. A modicum more effort upfront, but makes everything easier downstream. You can just run it directly, no need to set up a database server or anything like that (no idea if you can get it to run in an ipython notebook, I'll find out about native python options as that might be nicer)

In theory the hegroup one should work, but sorry I mislead the query will need tweaked, the hegroup sparql has a ton of other ontologies so it will need to be restricted to the specified named graph.

(Changing the URL shouldn't be necessary and in fact may break things - a SPARQL processor will not try and resolve these, they are just names that happen to be URLs)
